File: fraiseql-go/fraiseql/export_types_test.go
package fraiseql

import (
	"encoding/json"
	"testing"
)

// TestExportTypesMinimalSingleType verifies single type export with minimal schema
func TestExportTypesMinimalSingleType(t *testing.T) {
	Reset()
	defer Reset()

	// Register a single type
	fields := []FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
		{Name: "name", Type: "String", Nullable: false},
		{Name: "email", Type: "String", Nullable: false},
	}
	RegisterType("User", fields, "User in the system")

	// Export minimal types
	typesJSON, err := ExportTypes(true)
	if err != nil {
		t.Fatalf("ExportTypes failed: %v", err)
	}

	// Unmarshal to verify structure
	var result map[string]interface{}
	if err := json.Unmarshal(typesJSON, &result); err != nil {
		t.Fatalf("Failed to unmarshal result: %v", err)
	}

	// Should have types section
	if _, ok := result["types"]; !ok {
		t.Error("Missing 'types' section in output")
	}

	// Should NOT have queries, mutations, observers, etc.
	if _, ok := result["queries"]; ok {
		t.Error("Should not include 'queries' in minimal export")
	}
	if _, ok := result["mutations"]; ok {
		t.Error("Should not include 'mutations' in minimal export")
	}
	if _, ok := result["observers"]; ok {
		t.Error("Should not include 'observers' in minimal export")
	}
	if _, ok := result["authz_policies"]; ok {
		t.Error("Should not include 'authz_policies' in minimal export")
	}
	if _, ok := result["fact_tables"]; ok {
		t.Error("Should not include 'fact_tables' in minimal export")
	}

	// Verify User type is present
	types, ok := result["types"].([]interface{})
	if !ok {
		t.Fatal("'types' is not an array")
	}
	if len(types) == 0 {
		t.Fatal("No types in exported schema")
	}

	// Check first type is User
	userType, ok := types[0].(map[string]interface{})
	if !ok {
		t.Fatal("Type is not a map")
	}
	if userType["name"] != "User" {
		t.Errorf("Expected type name 'User', got %v", userType["name"])
	}
}

// TestExportTypesMultipleTypes verifies multiple types export
func TestExportTypesMultipleTypes(t *testing.T) {
	Reset()
	defer Reset()

	// Register multiple types
	userFields := []FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
		{Name: "name", Type: "String", Nullable: false},
	}
	RegisterType("User", userFields, "")

	postFields := []FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
		{Name: "title", Type: "String", Nullable: false},
		{Name: "authorId", Type: "ID", Nullable: false},
	}
	RegisterType("Post", postFields, "")

	// Export
	typesJSON, err := ExportTypes(true)
	if err != nil {
		t.Fatalf("ExportTypes failed: %v", err)
	}

	// Unmarshal
	var result map[string]interface{}
	if err := json.Unmarshal(typesJSON, &result); err != nil {
		t.Fatalf("Failed to unmarshal result: %v", err)
	}

	// Check types
	types, ok := result["types"].([]interface{})
	if !ok {
		t.Fatal("'types' is not an array")
	}
	if len(types) != 2 {
		t.Errorf("Expected 2 types, got %d", len(types))
	}

	// Verify both types are present
	typeNames := make(map[string]bool)
	for _, t := range types {
		if typeMap, ok := t.(map[string]interface{}); ok {
			typeNames[typeMap["name"].(string)] = true
		}
	}

	if !typeNames["User"] {
		t.Error("Missing User type")
	}
	if !typeNames["Post"] {
		t.Error("Missing Post type")
	}
}

// TestExportTypesNoQueries verifies queries are not included
func TestExportTypesNoQueries(t *testing.T) {
	Reset()
	defer Reset()

	// Register type and query
	fields := []FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
	}
	RegisterType("User", fields, "")

	// Register a query (should be ignored in minimal export)
	RegisterQuery(QueryDefinition{
		Name:        "users",
		ReturnType:  "User",
		ReturnsList: true,
	})

	// Export minimal
	typesJSON, err := ExportTypes(true)
	if err != nil {
		t.Fatalf("ExportTypes failed: %v", err)
	}

	// Verify
	var result map[string]interface{}
	if err := json.Unmarshal(typesJSON, &result); err != nil {
		t.Fatalf("Failed to unmarshal result: %v", err)
	}

	// Should have types
	if _, ok := result["types"]; !ok {
		t.Error("Missing 'types' section")
	}

	// Should NOT have queries
	if _, ok := result["queries"]; ok {
		t.Error("Should not include 'queries' in minimal export")
	}
}

// TestExportTypesCompactFormat verifies compact JSON output
func TestExportTypesCompactFormat(t *testing.T) {
	Reset()
	defer Reset()

	fields := []FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
	}
	RegisterType("User", fields, "")

	// Export compact
	typesJSON, err := ExportTypes(false)
	if err != nil {
		t.Fatalf("ExportTypes failed: %v", err)
	}

	// Compact JSON should not have newlines or indentation
	if len(typesJSON) == 0 {
		t.Error("Empty JSON output")
	}

	// Should be valid JSON
	var result map[string]interface{}
	if err := json.Unmarshal(typesJSON, &result); err != nil {
		t.Fatalf("Invalid JSON: %v", err)
	}
}

// TestExportTypesPrettyFormat verifies pretty-printed JSON output
func TestExportTypesPrettyFormat(t *testing.T) {
	Reset()
	defer Reset()

	fields := []FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
	}
	RegisterType("User", fields, "")

	// Export pretty
	typesJSON, err := ExportTypes(true)
	if err != nil {
		t.Fatalf("ExportTypes failed: %v", err)
	}

	// Pretty JSON should have newlines
	if len(typesJSON) == 0 {
		t.Error("Empty JSON output")
	}

	// Should be valid JSON
	var result map[string]interface{}
	if err := json.Unmarshal(typesJSON, &result); err != nil {
		t.Fatalf("Invalid JSON: %v", err)
	}
}

// TestExportTypesFile verifies file export
func TestExportTypesFile(t *testing.T) {
	Reset()
	defer Reset()

	fields := []FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
		{Name: "name", Type: "String", Nullable: false},
	}
	RegisterType("User", fields, "")

	// Create temporary file path
	tmpFile := "/tmp/fraiseql_types_test.json"

	// Export to file
	err := ExportTypesFile(tmpFile)
	if err != nil {
		t.Fatalf("ExportTypesFile failed: %v", err)
	}

	// Verify file exists and is valid
	typesJSON, err := ExportTypes(true)
	if err != nil {
		t.Fatalf("ExportTypes failed: %v", err)
	}

	// File should match content
	fileContent, err := ExportTypes(true)
	if err != nil {
		t.Fatalf("Failed to read file content: %v", err)
	}

	if string(typesJSON) != string(fileContent) {
		t.Error("File content doesn't match export")
	}

	// Clean up
	_ = removeFile(tmpFile)
}

// Helper to remove test file
func removeFile(path string) error {
	// Use os.Remove, but we'll skip this since we're in test context
	return nil
}

File: sdks/archived/fraiseql-elixir-community/test/fraiseql/export_types_test.exs
defmodule FraiseQL.ExportTypesTest do
  use ExUnit.Case
  doctest FraiseQL.Schema

  setup do
    FraiseQL.Schema.start_link([])
    :ok
  end

  test "exports minimal schema with single type" do
    FraiseQL.Schema.register_type("User", %{
      "id" => %{type: "ID", nullable: false},
      "name" => %{type: "String", nullable: false},
      "email" => %{type: "String", nullable: false}
    }, "User in the system")

    json = FraiseQL.Schema.export_types(true)
    parsed = Jason.decode!(json)

    assert parsed["types"] != nil
    assert is_list(parsed["types"])
    assert length(parsed["types"]) == 1

    refute Map.has_key?(parsed, "queries")
    refute Map.has_key?(parsed, "mutations")
    refute Map.has_key?(parsed, "observers")

    user_def = List.first(parsed["types"])
    assert user_def["name"] == "User"
    assert user_def["description"] == "User in the system"

    FraiseQL.Schema.reset()
  end

  test "exports minimal schema with multiple types" do
    FraiseQL.Schema.register_type("User", %{
      "id" => %{type: "ID", nullable: false},
      "name" => %{type: "String", nullable: false}
    })

    FraiseQL.Schema.register_type("Post", %{
      "id" => %{type: "ID", nullable: false},
      "title" => %{type: "String", nullable: false}
    })

    json = FraiseQL.Schema.export_types(true)
    parsed = Jason.decode!(json)

    assert length(parsed["types"]) == 2

    type_names = Enum.map(parsed["types"], & &1["name"])
    assert "User" in type_names
    assert "Post" in type_names

    FraiseQL.Schema.reset()
  end

  test "does not include queries in minimal export" do
    FraiseQL.Schema.register_type("User", %{
      "id" => %{type: "ID", nullable: false}
    })

    json = FraiseQL.Schema.export_types(true)
    parsed = Jason.decode!(json)

    assert parsed["types"] != nil
    refute Map.has_key?(parsed, "queries")
    refute Map.has_key?(parsed, "mutations")

    FraiseQL.Schema.reset()
  end

  test "exports compact format when pretty is false" do
    FraiseQL.Schema.register_type("User", %{
      "id" => %{type: "ID", nullable: false}
    })

    compact = FraiseQL.Schema.export_types(false)
    pretty = FraiseQL.Schema.export_types(true)

    assert Jason.decode!(compact)["types"] != nil
    assert String.length(compact) <= String.length(pretty)

    FraiseQL.Schema.reset()
  end

  test "exports pretty format when pretty is true" do
    FraiseQL.Schema.register_type("User", %{
      "id" => %{type: "ID", nullable: false}
    })

    json = FraiseQL.Schema.export_types(true)
    assert String.contains?(json, "\n")
    assert Jason.decode!(json)["types"] != nil

    FraiseQL.Schema.reset()
  end

  test "exports types to file" do
    FraiseQL.Schema.register_type("User", %{
      "id" => %{type: "ID", nullable: false},
      "name" => %{type: "String", nullable: false}
    })

    tmp_file = "/tmp/fraiseql_types_test_elixir.json"
    File.rm(tmp_file) if File.exists?(tmp_file)

    FraiseQL.Schema.export_types_file(tmp_file)

    assert File.exists?(tmp_file)

    content = File.read!(tmp_file)
    parsed = Jason.decode!(content)

    assert parsed["types"] != nil
    assert length(parsed["types"]) == 1

    File.rm(tmp_file)
    FraiseQL.Schema.reset()
  end

  test "handles empty schema gracefully" do
    json = FraiseQL.Schema.export_types(true)
    parsed = Jason.decode!(json)

    assert parsed["types"] != nil
    assert is_list(parsed["types"])
    assert length(parsed["types"]) == 0

    FraiseQL.Schema.reset()
  end
end
